File: Rishabh_Gupta/MinSpanTree.java
import java.util.Arrays;
import java.util.Scanner;
import java.util.PriorityQueue;
import java.util.Comparator;

public class MinSpanTree {
    public static int find(int[] parent, int i) {
        if (parent[i] != i) {
            parent[i] = find(parent, parent[i]);
        }
        return parent[i];
    }

    public static void union(int[] parent, int[] rank, int x, int y) {
        int xRoot = find(parent, x);
        int yRoot = find(parent, y);

        if (rank[xRoot] < rank[yRoot]) {
            parent[xRoot] = yRoot;
        } else if (rank[xRoot] > rank[yRoot]) {
            parent[yRoot] = xRoot;
        } else {
            parent[yRoot] = xRoot;
            rank[xRoot]++;
        }
    }

    public static int[][] kruskal(int[][] graph, int n) {
        int[] parent = new int[n];
        int[] rank = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }

        int[][] result = new int[n - 1][2];
        int resultIndex = 0;

        Arrays.sort(graph, (a, b) -> a[2] - b[2]);

        int i = 0;
        while (resultIndex < n - 1) {
            int[] edge = graph[i++];
            int x = find(parent, edge[0]);
            int y = find(parent, edge[1]);

            if (x != y) {
                result[resultIndex++] = edge;
                union(parent, rank, x, y);
            }
        }
        return result;
    }

    public static int[][] prim(int[][] graph, int n) {
        boolean[] inMST = new boolean[n];
        int[][] result = new int[n - 1][2];
        int resultIndex = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[2]));
        pq.add(new int[]{0, 0, 0});

        while (!pq.isEmpty() && resultIndex < n - 1) {
            int[] edge = pq.poll();
            int u = edge[0];
            int v = edge[1];
            int weight = edge[2];

            if (inMST[v]) continue;

            inMST[v] = true;
            if (u != v) {
                result[resultIndex++] = new int[]{u, v, weight};
            }

            for (int[] nextEdge : graph) {
                if (nextEdge[0] == v && !inMST[nextEdge[1]]) {
                    pq.add(new int[]{v, nextEdge[1], nextEdge[2]});
                } else if (nextEdge[1] == v && !inMST[nextEdge[0]]) {
                    pq.add(new int[]{v, nextEdge[0], nextEdge[2]});
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int[][] graph = {{0, 1, 2}, {1, 2, 3}, {0, 3, 6}, {1, 3, 8}, {1, 4, 5}, {2, 4, 7}};
        int n = 5;

        System.out.println("Choose algorithm: 1 for Kruskal, 2 for Prim");
        int choice = scanner.nextInt();

        int[][] mst;
        if (choice == 1) {
            mst = kruskal(graph, n);
        } else {
            mst = prim(graph, n);
        }

        System.out.println("Minimum Spanning Tree:");
        for (int[] edge : mst) {
            System.out.println(Arrays.toString(edge));
            scanner.close();
        }
    }
}

File: Venkatesh/Advance-Graph/Kruskal/MinSpanTree.c
#include <stdio.h>
#include <stdlib.h>

struct Edge {
    int src, dest, weight;
};

struct Subset {
    int parent;
    int rank;
};

int find(struct Subset subsets[], int i) {
    if (subsets[i].parent != i)
        subsets[i].parent = find(subsets, subsets[i].parent);
    return subsets[i].parent;
}

void unionSets(struct Subset subsets[], int x, int y) {
    int xroot = find(subsets, x);
    int yroot = find(subsets, y);

    if (subsets[xroot].rank < subsets[yroot].rank)
        subsets[xroot].parent = yroot;
    else if (subsets[xroot].rank > subsets[yroot].rank)
        subsets[yroot].parent = xroot;
    else {
        subsets[yroot].parent = xroot;
        subsets[xroot].rank++;
    }
}

int compareEdges(const void* a, const void* b) {
    struct Edge* edge1 = (struct Edge*)a;
    struct Edge* edge2 = (struct Edge*)b;
    return edge1->weight - edge2->weight;
}

void kruskal(int graph[][3], int n) {
    struct Edge result[n];
    int e = 0;
    int i = 0;

    qsort(graph, n, sizeof(struct Edge), compareEdges);

    struct Subset* subsets = (struct Subset*)malloc(n * sizeof(struct Subset));

    for (int v = 0; v < n; v++) {
        subsets[v].parent = v;
        subsets[v].rank = 0;
    }

    while (e < n - 1 && i < n) {
        struct Edge nextEdge = *((struct Edge*)(&graph[i++]));

        int x = find(subsets, nextEdge.src);
        int y = find(subsets, nextEdge.dest);

        if (x != y) {
            result[e++] = nextEdge;
            unionSets(subsets, x, y);
        }
    }

    for (i = 0; i < e; i++) {
        printf("[%d, %d, %d]", result[i].src, result[i].dest, result[i].weight);
        if (i < e - 1) {
            printf(", ");
        }
    }
    printf("]\n");

    free(subsets);
}

void main() {
    int graph[][3] = {{0, 1, 2}, {1, 2, 3}, {0, 3, 6}, {1, 3, 8}, {1, 4, 5}, {2, 4, 7}};
    int n = 5;

    kruskal(graph, n);
}